service/history/workflow/update/update_test.go

```diff
@@ -543,3 +543,43 @@ func TestRollbackCompletion(t *testing.T) {
 		require.NoError(t, err, "update should be back in Accepted state")
 	})
 }
+
+func TestRejectionWithAcceptanceWaiter(t *testing.T) {
+	t.Parallel()
+	var (
+		ctx      = context.Background()
+		store    = mockEventStore{Controller: effect.Immediate(ctx)}
+		updateID = t.Name() + "-update-id"
+		upd      = update.New(updateID, ignoreCompletion)
+		req      = updatepb.Request{
+			Meta:  &updatepb.Meta{UpdateId: updateID},
+			Input: &updatepb.Input{Name: "not_empty"},
+		}
+		rej = updatepb.Rejection{
+			RejectedRequestMessageId: "not_empty",
+			RejectedRequest:          &req,
+			Failure: &failurepb.Failure{
+				Message: "intentional falure from " + t.Name(),
+			},
+		}
+	)
+	ch := make(chan any, 1)
+	go func() {
+		fail, err := upd.WaitAccepted(ctx)
+		if err != nil {
+			ch <- err
+			return
+		}
+		ch <- fail
+	}()
+	t.Log("give 5ms for the goro to get to the Future.Get call in WaitAccepted")
+	time.Sleep(5 * time.Millisecond)
```

Comment:
Why do we need this if we're already doing a blocking read on the channel without a timeout below? If possible, please remove this

Reply:
It's not required for the test to pass, just for the test to cover the branch that I want. I want the goro to get past the first check to Update.outcome.Ready in Update. This unit test is explicitly to cover a concurrent condition where there's already a waiter at line [162](https://github.com/temporalio/temporal/pull/4311/files#diff-89d9f00e7fa3af35a4d4a32d21edb961ad3810e10dbdbb2b508b9ce326e8104fR162) of Update.go when the rejection is received. To do that, I have to give it time to get there.

Reply:
`runtime.Gosched()` ?

Reply:
Both that approach and the time sleep are still technically data races. I'd prefer we use a [test seam](https://blog.sasworkshops.com/unit-testing-and-seams/) (for lack of a better term) here. For example, decorate the `.Get` method of the `accepted` future, and wait for that to be called.